### ISSUE TITLE
WebSockets and HTTP/2

### ISSUE THREAD

**Issue by kjetilk** [solid/solid-spec#123]
I think we need to review our use of Websockets. 

Now, Websockets are supported in a [standalone package](https://github.com/solid/node-solid-ws), which has not received much attention. The use cases that motivates Websocket, could to a great extent be done using [HTTP/2 and SSE](https://www.infoq.com/articles/websocket-and-http2-coexist), even though HTTP/2 doesn't entirely replace Websockets.

We should discuss whether we should bring such use cases under the HTTP umbrella, support HTTP/2 for Solid and integrate SSE into the platform.

**Comment by elf-pavlik** [comment]
:+1: I suggested SSE in #11 

Also in https://github.com/solid/solid/issues/116 I asked what advantage globing gives in HTTP/2 world

**Comment by kjetilk** [comment]
Just a short note on SSE, since I just read [the spec](https://www.w3.org/TR/2015/REC-eventsource-20150203/), and I didn't quite like what I read. SSE does not maintain orthogonality is a principle, and as a result, it defines both an API and a media type. 

The media type does not seem to make much sense from a Solid perspective, but if we had an RDF-based format for the same thing, which I think is something the academic community has been working on (citation needed, I don't have it), it would make sense.

I ended up concluding that SSE didn't make sense for Solid, but an extension where `EventSource` is specified separately from the media type would. So, perhaps that is something we should look into, even though that's quite a lot larger.

**Comment by gobengo** [comment]
> The media type does not seem to make much sense from a Solid perspective, but if we had an RDF-based format for the same thing, which I think is something the academic community has been working on (citation needed, I don't have it), it would make sense.

I don't understand your critique. RDF is a data model, right? Not a data serialization. How is this text/event-stream document not sufficiently "RDF-based"?:

```
event: https://www.w3.org/ns/activitystreams#Create
data: {"@context":"https://www.w3.org/ns/activitystreams","summary":"Sally created a note","type":"Create","actor":{"type":"Person","name":"Sally"},"object":{"type":"Note","name":"A Simple Note","content":"This is a simple note"}}

event: https://www.w3.org/ns/activitystreams#Update
data: "{"@context":"https://www.w3.org/ns/activitystreams","summary":"Sally updated her note","type":"Update","actor":{"type":"Person","name":"Sally"},"object":"http://example.org/notes/1"}"
``

> I ended up concluding that SSE didn't make sense for Solid, but an extension where EventSource is specified separately from the media type would.

but then it wouldn't work in all [these browsers](https://caniuse.com/#search=EventSource) implementations... What would be the point?

It seems reeeeaaally inappropriate to dismiss these widely implemented protocols and user-agent API implementations just because you disagree with an editorial decision on how the spec was authored [10 years ago](https://www.w3.org/TR/2009/WD-eventsource-20090423/).

HTTP/2 is cool but 20 years of user-agents won't support it. SSE is over HTTP (unlike WebSocket) and can be polyfilled all the way back to IE8. I don't think WebSocket or HTTP2 can do the same without resorting to insecure plugin things like Flash.

**Comment by elf-pavlik** [comment]
> The media type does not seem to make much sense from a Solid perspective, but if we had an RDF-based format for the same thing, which I think is something the academic community has been working on (citation needed, I don't have it), it would make sense.

How does SSE differ in this regard from current Solid Websockets API?

https://github.com/solid/solid-spec/blob/master/api-websockets.md

> If a change occurs and the client is subscribed to that resource, it will receive a WebSocket message composed of the keyword `pub`, followed by an empty space and the URI of the resource that has changed:
>
>`pub https://example.org/data/test`

I don't see this format having anything to do with RDF.

**Comment by kjetilk** [comment]
> I don't understand your critique. RDF is a data model, right? 

Indeed, and I wrote "RDF-based" for exactly that reason.

>Not a data serialization. How is this text/event-stream document not sufficiently "RDF-based"?:

Because, the media type is then given as "text/event-stream", and not as an RDF serialization.
 
> ```
> event: https://www.w3.org/ns/activitystreams#Create
> data: {"@context":"https://www.w3.org/ns/activitystreams","summary":"Sally created a note","type":"Create","actor":{"type":"Person","name":"Sally"},"object":{"type":"Note","name":"A Simple Note","content":"This is a simple note"}}
> 
> event: https://www.w3.org/ns/activitystreams#Update
> data: "{"@context":"https://www.w3.org/ns/activitystreams","summary":"Sally updated her note","type":"Update","actor":{"type":"Person","name":"Sally"},"object":"http://example.org/notes/1"}"
> ```

Right, so you have to make assumptions about what the serialization is without having any metadata to rely on. That's not nice, but yeah, an assumption that it is JSON is likely to go a long way.

> It seems reeeeaaally inappropriate to dismiss these widely implemented protocols and user-agent API implementations just because you disagree with an editorial decision on how the spec was authored [10 years ago](https://www.w3.org/TR/2009/WD-eventsource-20090423/).

It is more than an editorial decision. It is a violation of one of the core [architectural principles](https://www.w3.org/TR/webarch/Overview.html#orthogonal-specs) of the Web dating back to its birth. If HTML and HTTP had been in the same spec, it would probably never have taken off.

But hey, it is just about making something that works. If stuffing some JSON-LD in there will keep people happy and productive, I'm not going to be in the way, I'm just saying this could have been done better, and possibly, we should.

**Comment by gobengo** [comment]
>Because, the media type is then given as "text/event-stream", and not as an RDF serialization..

Is there a source of truth for the blessed set of "RDF Serializations" you can link me to, or is that assertion just your opinion? If I define "text/event-stream+rdf" and add it to [this wiki page](https://www.w3.org/wiki/RdfSyntax), would that make it one?

> Right, so you have to make assumptions about what the serialization is without having any metadata to rely on. That's not nice, but yeah, an assumption that it is JSON is likely to go a long way.

I thought that might be the nature of your concern. What if the aforementioned text/event-stream document had come from a request with headers:

```
Content-Type: text/event-stream;
Sec-Event-Stream-Protocol: application/ld+json
```

(or, to keep it as a parameter in the same IANA media type registration, add a media type parameter like `Content-Type: text/event-stream; dataMediaType=application/ld+json`

(Where `Sec-Event-Stream-Protocol` is like the [WebSocket Subprotocol Header](https://tools.ietf.org/html/rfc6455#section-1.9))

A corollary is that this could also be a thing

Headers
```
Content-Type: text/turtle;
Sec-Event-Stream-Protocol: text/turtle
```
(or `Content-Type: text/event-stream; dataMediaType=text/turtle`)

Response
```
event: https://www.w3.org/ns/activitystreams#Create
data:  <http://example.org/person/Mark_Twain>
data:    <http://example.org/relation/author>
data:    <http://example.org/books/Huckleberry_Finn> .
```

It seems like this would bring back the orthogonality you value by making `text/event-stream` ['generic'](https://en.wikipedia.org/wiki/Generic_programming) like HTTP itself. But, as always, I could be misunderstanding.

**Comment by kjetilk** [comment]
> > Because, the media type is then given as "text/event-stream", and not as an RDF serialization..
> 
> Is there a source of truth for the blessed set of "RDF Serializations" you can link me to, or is that assertion just your opinion?

I am aware of no blessed set of RDF serializations, but I do know what the media types of those that have been implemented are. Generally, there is an open world assumption around these things. So, if you assert that "text/event-stream" is one, please show how it conforms to the RDF abstract syntax.

> If I define "text/event-stream+rdf" and add it to [this wiki page](https://www.w3.org/wiki/RdfSyntax), would that make it one?

No, since RDF can be serialized in many ways. So, you could in principle define "text/event-stream+turtle", etc.


> > Right, so you have to make assumptions about what the serialization is without having any metadata to rely on. That's not nice, but yeah, an assumption that it is JSON is likely to go a long way.
> 
> I thought that might be the nature of your concern. What if the aforementioned text/event-stream document had come from a request with headers:
> 
> ```
> Content-Type: text/event-stream;
> Sec-Event-Stream-Protocol: application/ld+json
> ```

Yeah, that could be a reasonable direction to go in.
 
> (or, to keep it as a parameter in the same IANA media type registration, add a media type parameter like `Content-Type: text/event-stream; dataMediaType=application/ld+json`

Yup, that too could work.

> A corollary is that this could also be a thing
> 
> Headers
> 
> ```
> Content-Type: text/turtle;
> Sec-Event-Stream-Protocol: text/turtle
> ```

Indeed, but then you'd need to define media types, protocols and APIs orthogonally.
 
> It seems like this would bring back the orthogonality you value by making `text/event-stream` ['generic'](https://en.wikipedia.org/wiki/Generic_programming) like HTTP itself. But, as always, I could be misunderstanding.

Indeed, it could be done.

**Comment by elf-pavlik** [comment]
I think https://www.w3.org/TR/rdf11-new/#section-serializations lists all the official RDF1.1 serializations. 

I also stumbled upon https://profilenegotiation.github.io/I-D-Accept--Schema/I-D-accept-schema which I understand addresses different problem but maybe based on experience in that work @RubenVerborgh has some suggestions how `text/event-stream` could work with varous RDF serializations.

**Comment by kjetilk** [comment]
> I think https://www.w3.org/TR/rdf11-new/#section-serializations lists all the official RDF1.1 serializations.
>

Yeah, but there is a bunch of binary serializations too, and I think people have been working on serializations of event streams specifically, so I would prefer to assume an open world :-)
